src/terraformpy/objects.py

```diff
@@ -42,15 +42,40 @@ class DuplicateKey(str):
     # Preserving usage order will help plan output remain consistent across invocations
     next_hash_per_str = collections.defaultdict(lambda: 0)
 
-    def __init__(self, key=''):
-        super(DuplicateKey, self).__init__(key)
+    def __new__(cls, key):
+        inst = super(DuplicateKey, cls).__new__(cls, key)
 
-        self.hash = DuplicateKey.next_hash_per_str[key]
-        DuplicateKey.next_hash_per_str[key] = self.hash + 1
+        inst.hash = DuplicateKey.next_hash_per_str[key]
+        DuplicateKey.next_hash_per_str[key] = inst.hash + 1
+
+        return inst
 
     def __hash__(self):
         return self.hash
 
+    def __eq__(self, other):
+        return self.__class__ == other.__class__ and self.hash == other.hash
```

Comment:
Per [the Python 3 data model docs on `__hash__`](https://docs.python.org/3/reference/datamodel.html#object.__hash__):

> If a class does not define an `__eq__()` method it should not define a `__hash__()` operation either

Without `__eq__` and the other rich comparison methods the py3 tests were failing with:

```
______________________________ test_duplicate_key ______________________________

    def test_duplicate_key():
        # ordering should ensure it's always <firstCreatedKey>, <secondCreatedKey>
        # so let's ensure that works, even if say the value names are backwards-sorted
        key2 = DuplicateKey("mysql")
        key1 = DuplicateKey("mysql")
>       encoded = json.dumps({key1: {"user": "wyatt1"}, key2: {"user": "wyatt2"}}, sort_keys=True)

tests/test_objects.py:183: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/__init__.py:238: in dumps
    **kw).encode(obj)
/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py:199: in encode
    chunks = self.iterencode(o, _one_shot=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <json.encoder.JSONEncoder object at 0x10974b150>, o = {'mysql': {'user': 'wyatt1'}, 'mysql': {'user': 'wyatt2'}}, _one_shot = True

    def iterencode(self, o, _one_shot=False):

<snip...>

>       return _iterencode(o, 0)
E       TypeError: '<' not supported between instances of 'dict' and 'dict'
```